manticore/native/models.py

```diff
@@ -140,3 +143,124 @@ def strlen(state, s):
+
+    :param byte: byte read from memory to be examined
+    :param constrs: state constraints
+    :return: whether a given byte is NULL or can be NULL
+    """
+    if issymbolic(byte):
+        return Z3Solver.instance().can_be_true(constrs, byte == 0)
+    else:
+        return byte == 0
+
+
+def strcpy(state: State, dst: Union[int, BitVec], src: Union[int, BitVec]) -> Union[int, BitVec]:
+    """
+    strcpy symbolic model
+
+    Algorithm: Copy every byte from the src to dst until finding a byte that can be or is NULL.
+    If the byte is NULL or is constrained to only the NULL value, append the NULL value to dst
+    and return. If the value can be NULL or another value write an `Expression` for every following
+    byte that sets a value to the src or dst byte according to the preceding bytes until a NULL
+    byte or effectively NULL byte is found.
+
+    :param state: current program state
+    :param dst: destination string address
+    :param src: source string address
+    :return: pointer to the dst
+    """
+    if issymbolic(src):
+        raise ConcretizeArgument(state.cpu, 1)
+
+    if issymbolic(dst):
+        raise ConcretizeArgument(state.cpu, 2)
+
+    cpu = state.cpu
+    constrs = state.constraints
+    ret = dst
+
+    # Copy until '\000' is reached or symbolic memory that can be '\000'
+    c = cpu.read_int(src, 8)
+    while cant_be_NULL(c, constrs):
+        cpu.write_int(dst, c, 8)
+        src += 1
+        dst += 1
+        c = cpu.read_int(src, 8)
+
+    # If the byte is symbolic and constrained to '\000', or is concrete and '\000', write concrete val and return
+    if is_definitely_NULL(c, constrs):
+        cpu.write_int(dst, 0, 8)
+        return ret
+
+    offset = 0
+    zeros: Deque[int] = deque([])
+    src_val = cpu.read_int(src, 8)
+    while not is_definitely_NULL(src_val, constrs):
+        if can_be_NULL(c, constrs):
+            # If a byte can be NULL set the src_val for NULL, build the ITE, & add to the list of nulls
+            src_val = ITEBV(8, src_val != 0, src_val, 0)
+            _build_ITE(zeros, cpu, src, dst, offset, src_val)
+            zeros.appendleft(offset)
+        else:
+            # If it can't be NULL just build the ITE
+            _build_ITE(zeros, cpu, src, dst, offset, src_val)
+        offset += 1
+        src_val = cpu.read_int(src + offset, 8)
+
+    # Build ITE Tree for NULL byte
+    src_val = 0
+    _build_ITE(zeros, cpu, src, dst, offset, src_val)
+
+    return ret
+
+
+def _build_ITE(
+    zeros: Deque[int],
+    cpu: Cpu,
+    src: Union[int, BitVec],
+    dst: Union[int, BitVec],
+    offset: int,
+    src_val: Union[int, BitVec],
+) -> None:
+    """
+    Builds ITE tree for each symbolic dst byte
+    """
+    dst_val = cpu.read_int(dst + offset, 8)
+    for zero in zeros:
+        c = cpu.read_int(src + zero, 8)
+        src_val = ITEBV(8, c != 0, src_val, dst_val)
+    cpu.write_int(dst + offset, src_val, 8)
```

Comment:
Since `_build_ITE` is a helper function, it makes me a little bit uncomfortable that it has side effects. Instead, can we return `src_val` and pass it to `cpu.write_int` in `strcpy`?

Reply:
Will do :)